File: leetcode/Arrays/leetcode-0054/54.java
class Solution {
    public List<Integer> spiralOrder(int[][] matrix) {
        List<Integer> res = new ArrayList<>();
        if (matrix == null || matrix.length == 0 || matrix[0].length == 0) {
            return res;
        }
        int row = matrix.length - 1, col = matrix[0].length - 1;
        int r = 0, c = 0;
        while (true) {
            for (int i = c; i <= col; i++) {
                res.add(matrix[r][i]);
            }
            if (++r > row) {
                break;
            }

            for (int i = r; i <= row; i++) {
                res.add(matrix[i][col]);
            }
            if (c > --col) {
                break;
            }

            for (int i = col; i >= c; i--) {
                res.add(matrix[row][i]);
            }
            if (r > --row) {
                break;
            }

            for (int i = row; i >= r; i--) {
                res.add(matrix[i][c]);
            }
            if (++c > col) {
                break;
            }
        }
        return res;
    }
}

File: leetcode/Arrays/leetcode-0054/54.go
package arrays

func spiralOrder(matrix [][]int) []int {
	res := []int{}
	if matrix == nil || len(matrix) == 0 || len(matrix[0]) == 0 {
		return res
	}

	r, c, row, col := 0, 0, len(matrix)-1, len(matrix[0])-1
	for {
		for i := c; i <= col; i++ {
			res = append(res, matrix[r][i])
		}
		r++
		if r > row {
			break
		}

		for i := r; i <= row; i++ {
			res = append(res, matrix[i][col])
		}
		col--
		if c > col {
			break
		}

		for i := col; i >= c; i-- {
			res = append(res, matrix[row][i])
		}
		row--
		if r > row {
			break
		}

		for i := row; i >= r; i-- {
			res = append(res, matrix[i][c])
		}
		c++
		if c > col {
			break
		}
	}
	return res
}
